common/oauthTokenManager.go

```diff
@@ -142,6 +149,203 @@ func (uotm *UserOAuthTokenManager) MSILogin(ctx context.Context, identityInfo Id
+	}
+
+	if applicationID == "" {
+		return nil, fmt.Errorf("please supply your OWN application ID")
+	}
+
+	oAuthTokenInfo := OAuthTokenInfo{
+		Tenant:                  tenantID,
+		ActiveDirectoryEndpoint: activeDirectoryEndpoint,
+	}
+
+	oauthConfig, err := adal.NewOAuthConfig(activeDirectoryEndpoint, tenantID)
+	if err != nil {
+		return nil, err
+	}
+
+	spt, err := adal.NewServicePrincipalToken(
+		*oauthConfig,
+		applicationID,
+		secret,
+		Resource,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	err = spt.Refresh()
+	if err != nil {
+		return nil, err
+	}
+
+	oAuthTokenInfo.Token = spt.Token()
+	oAuthTokenInfo.ApplicationID = applicationID
+	oAuthTokenInfo.ServicePrincipalName = true
+	oAuthTokenInfo.SPNInfo = SPNInfo{
+		Secret:   secret,
+		CertPath: "",
+	}
+
+	return &oAuthTokenInfo, nil
+}
+
+//SecretLogin is a UOTM shell for secretLoginNoUOTM.
+func (uotm *UserOAuthTokenManager) SecretLogin(tenantID, activeDirectoryEndpoint, secret, applicationID string, persist bool) (*OAuthTokenInfo, error) {
+	oAuthTokenInfo, err := secretLoginNoUOTM(tenantID, activeDirectoryEndpoint, secret, applicationID)
+
+	if persist {
+		err = uotm.credCache.SaveToken(*oAuthTokenInfo)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return oAuthTokenInfo, nil
+}
+
+//GetNewTokenFromSecret is a refresh shell for secretLoginNoUOTM
+func (credInfo *OAuthTokenInfo) GetNewTokenFromSecret(ctx context.Context) (*adal.Token, error) {
+	tokeninfo, err := secretLoginNoUOTM(credInfo.Tenant, credInfo.ActiveDirectoryEndpoint, credInfo.SPNInfo.Secret, credInfo.ApplicationID)
+
+	if err != nil {
+		return nil, err
+	} else {
+		return &tokeninfo.Token, nil
+	}
+}
+
+func certLoginNoUOTM(tenantID, activeDirectoryEndpoint, certPath, certPass, applicationID string) (*OAuthTokenInfo, error) {
+	if tenantID == "" {
+		tenantID = DefaultTenantID
+	}
+
+	if activeDirectoryEndpoint == "" {
+		activeDirectoryEndpoint = DefaultActiveDirectoryEndpoint
+	}
+
+	if applicationID == "" {
+		return nil, fmt.Errorf("please supply your OWN application ID")
+	}
+
+	oAuthTokenInfo := OAuthTokenInfo{
+		Tenant:                  tenantID,
+		ActiveDirectoryEndpoint: activeDirectoryEndpoint,
+	}
+
+	oauthConfig, err := adal.NewOAuthConfig(activeDirectoryEndpoint, tenantID)
+	if err != nil {
+		return nil, err
+	}
+
+	certData, err := ioutil.ReadFile(certPath)
+	if err != nil {
+		return nil, err
+	}
+
+	var pk interface{}
+	var cert *x509.Certificate
+
+	if path.Ext(certPath) == ".pfx" {
+		pk, cert, err = pkcs12.Decode(certData, certPass)
+		if err != nil {
+			return nil, err
+		}
+	} else if path.Ext(certPath) == ".pem" {
+		block, rest := pem.Decode(certData)
+
+		for len(rest) != 0 || pk == nil || cert == nil {
+			switch block.Type {
+			case "PRIVATE KEY":
+				pk, err = x509.ParsePKCS8PrivateKey(block.Bytes)
+
+				if err != nil {
+					return nil, fmt.Errorf("private key has invalid format")
+				}
+			case "CERTIFICATE":
+				cert, err = x509.ParseCertificate(block.Bytes)
```

Comment:
Ah! Good catch. You also made me wonder if PEM files can contain multiple chains, and I'm not certain that they can at all.

I'm *guessing* it would be quite hard to implement multiple cert chains with this spec, but I think on the off-chance I'm going to add in a prompt following sync's example.

I'll make it so that it stores that alongside other SPN info.

Reply:
I still left the entirety of the original functionality as well, so it's not too hard to get rid of if we decide _not_ to support this use case.

Reply:
>so it's not too hard to get rid of if we decide not to support this use case.

Yeah, I'm leaning towards the view that we should _not_ support the multiple chain scenario.  I.e. just assume there is only one chain (or better, just throw a descriptive error if there is more than one cert in the PEM where isCA == false).

Why not support it?  Because:
* it adds complexity (to both code and testing)
* it makes SPA interactive - when one of the key goals of it is that it should be a good solution for cases where everything is scripted (no human interaction)
* the above don't seem justified in a case where we are already doubtful that it will ever come up. (I.e. we think it probably won't)

Reply:
Can you ping me when its removed pls?